File: OlderCodes/Ccodes/eqheat.c
/* From: "A SURVEY OF COMPUTATIONAL PHYSICS" 
   by RH Landau, MJ Paez, and CC BORDEIANU 
   Copyright Princeton University Press, Princeton, 2008.
   Electronic Materials copyright: R Landau, Oregon State Univ, 2008;
   MJ Paez, Univ Antioquia, 2008; and CC BORDEIANU, Univ Bucharest, 2008.
   Support by National Science Foundation                              
   */
//  eqheat.c: Solution of heat equation using with finite differences

#include <stdlib.h> 
#include <stdio.h>
#define size 101                                         // grid size
#define max 30000                                       // iterations
#define thc 0.12                              // thermal conductivity
#define sph 0.113                                    // specific heat
#define rho 7.8                                            // density

main()  { 
	
   int i,j;
   double cons, u[101][2];     
	 
   FILE *output;                           // save data in eqheat.dat 
   output= fopen("eqheat.dat","w");
	                                           // t = 0 points are 100C
   for (i = 0; i<size; i++) u[i][0] = 100.;      
	                                            // except the endpoints
   for (j = 0; j<2; j++) {                          
     u[0][j]     = 0.;
     u[size-1][j]= 0.;
   } 
   cons = thc/(sph*rho);                        // material constants
	                                         // loop over max timesteps
   for (i = 1; i<= max; i++) {                  
		                                               // loop over space
     for (j = 1; j<(size-1); j++)                       
      {u[j][1]= u[j][0]+cons*(u[j+1][0]+u[j-1][0]-2.0*u[j][0]); }
		                                         // save every 1000 steps
      if ((i%1000 == 0) || (i == 1))  {   
        for (j = 0 ; j<size; j++) fprintf(output, "%f\n", u[j][1]);
        fprintf(output, "\n");              // empty line for gnuplot
      }
      for (j = 0; j<size; j++) u[j][0] = u[j][1];       // new to old
   }
   printf("data stored in eqheat.dat\n");
   fclose(output);
}      
   

File: OlderCodes/Fortran77codes/eqheat.f
c  From: "A SURVEY OF COMPUTATIONAL PHYSICS" 
c  by RH Landau, MJ Paez, and CC BORDEIANU 
c  Copyright Princeton University Press, Princeton, 2008.
c  Electronic Materials copyright: R Landau, Oregon State Univ, 2008;
c  MJ Paez, Univ Antioquia, 2008; and CC BORDEIANU, Univ Bucharest, 2008.
c  Support by National Science Foundation                             
c                                                              
c  eqheat.f: Solution of heat equation using with finite differences                                                 
c            Output data is saved in 3D grid format used by gnuplot 

	    Program heat
			
        Implicit None       
      	Double Precision cons, ro, sph, thk, u(101,2)
      	Integer i, k, max
				
      	Open(9,file = 'eqheat.dat',status = 'Unknown')
c            specific heat, thermal conductivity and density for iron
       	sph = 0.113
	      thk = 0.12
	      ro = 7.8
        cons = thk/(sph*ro)
c                                                number of iterations
    	  max = 30000
c                                    At t = 0 all points are at 100 C
      	Do 10 i = 1,100
          u(i,1) = 100.0
 10   	Continue
c                                                      endpoints zero
      	Do 20 i = 1,2
          u(1,i) = 0.0
          u(101,i) = 0.0
 20     Continue
c 			                                    start solution, time loop
      	Do 100 k = 1,max
c                               loop over space, endpoints stay fixed
          Do 30 i = 2,100
            u(i,2) = u(i,1) + cons*(u(i+1,1) + u(i-1,1)-2*u(i,1))
 30       Continue
c                                        output every 1000 time steps
          If ( (Mod(k,1000) .eq. 0)  .or.  (k .eq. 1) ) Then
            Do 40 i = 1,101,2
              Write(9,22)u(i,2)
 40         Continue
            Write (9,22)
          EndIf
c                                                          new -> old
          Do 50 i = 2,100
            u(i,1) = u(i,2)
 50       Continue
 100  	Continue
 22   	Format (f10.6)
 	      Close(9)
      	Stop 'data saved in eqheat.dat'
      End
